smolyay/samples.py

```diff
@@ -0,0 +1,207 @@
+import abc
+import sys
+import numpy
+
+
+class UnidimensionalPointSet(abc.ABC):
+    """Set of unidimensional points
+
+    A set of unique unidimensional points to be used as the sampling points
+    in a tensor product grid or sparse grid. These unidimensional points are
+    associated with certain polynomial families, and the combination of
+    unique points and polynomial families are the basis for different types
+    of quadrature rules that are used in numerical integration and in
+    approximation.
+    The natural domain of the points specified by :param:``natural_domain`` is
+    the domain in which all the possible points in the set lies within.
+    ``number_points`` is the number of points that the UnidimensionalPointSet
+    is expected to store.
+    ``domain`` is the domain of the points generated by the point set.
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @property
+    @abc.abstractmethod
+    def domain(self):
+        """list: Domain the sample points come from."""
+        pass
+
+    @property
+    def points(self):
+        """list: Points stored by the set."""
+        return self._points
+
+    @abc.abstractmethod
+    def generate_points(self, num_points):
+        r"""Generating the points
+
+        An abstract method for generating the points stored by the point set.
+        Called when the number of points requested is larger than the number
+        of points available.
+
+        Parameters
+        ----------
+        num_points : int
+            The number of points to generate
+
+        Returns
+        -------
+        self
+            The UnidimensionalPointSet
+        """
+        pass
+
+
+class ClenshawCurtisPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Clenshaw Curtis sampling
+
+    The :attr:`points` for this interpolation scheme are the extrema of the
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def domain(self):
+        """list: Domain the sample points come from."""
+        return [-1, 1]
+
+    def generate_points(self, num_points):
+        r"""Generating the points
+
+        Parameters
+        ----------
+        num_points : int
+            The number of points to generate
+
+        Returns
+        -------
+        self
+            The UnidimensionalPointSet
+        """
+        if num_points == 0:
+            self._points = []
+        degree = num_points - 1
+        if degree > 0:
+            self._points = list(
+                -numpy.cos(numpy.pi * numpy.linspace(0, degree, degree + 1) / degree)
+            )
+        else:
+            self._points = [0]
+        return self
+
+
+class NestedClenshawCurtisPointSet(ClenshawCurtisPointSet):
```

Comment:
This class should have its own constructor that takes something like the `max_level`
```py
def __init__(self, max_level):
    self.max_level = max_level
```

Reply:
Should degree and max_level have getter and setter methods?

Reply:
Yes, since we will be caching the values of `points`, you need to add a property setter for any attributes that invalidate the cache if their value changes.

Reply:
For the Trigonometric point set, is degree the most appropriate name for that variable? While `degree` makes sense for the Clenshaw Curtis point set since it's the degree of the Chebyshev polynomial, what is the "degree" in the context of the Trigonometric point set?

Reply:
No, probably not. You can choose an appropriate name and use it in that class. It does not need to be the same as the one used in Clenshaw-Curtis points

Reply:
I don't think number of points makes sense for the trigonometric points because they should be the roots of a certain frequency trig function. So, maybe `frequency` makes sense here?

And just to emphasize again: specific constructors for derived classes do not need to be the same. You should choose terminology that makes sense based on what they represent.

Reply:
Derived classes don't need to be the same as each other, but I feel it's prudent to point out that some amount of consistency between them is more convenient. When these classes are used by a multidim point set, like the one for creating tensor grids, that class isn't going to know what kind of UnidimensionalPointSet it's using nor should it. It shouldn't have to run different code depending on what child class it's using, so there should be some level of consistency.

And on a wider scale, the code using and initializing the `UnidimensionalPointSet` to make grids shouldn't have to care that much about what kind of `UnidimensionalPointSet` it's using either, which makes it easier to reuse code and predict how changes will affect the grids it generates.

As it stands now, the `ClenshawCurtisPointSet` when initialized will create a list of points with 1 more element than the number it was initialized with. The `TrigonometricPointSet` when initialized will create a list of points with as many elements as the number it was initialized with (with 0 being the exception). A person unfamiliar with this code and its inner workings may find this confusing and not understand why their tensor grids are generating with more or less points than they expect. You can't just switch the class type and expect to get a tensor grid with the exact same number of points unless you remember to change the input. 

And I imagine that would be very frustrating, especially since it would be trivial to make both classes give the same number of points as the input parameter.

It's the whole reason why `number_points` was a property of the base class when I first made the pull request. Even if you knew nothing about the deriving classes' inner workings, you would always know how many points you would get.

Reply:
> Derived classes don't need to be the same as each other, but I feel it's prudent to point out that some amount of consistency between them is more convenient. ... And on a wider scale, the code using and initializing the UnidimensionalPointSet to make grids shouldn't have to care that much about what kind of UnidimensionalPointSet it's using either

Yes, this is the whole idea of duck typing and defining an abstract base class with a common interface. The minimum attributes and methods that all concrete implementations need should be in the abstract base class, but this **does not** mean that their constructors have to be the same or that deriving classes cannot have attributes that are not in the base class. This is a fundamental concept in object-oriented program.

> You can't just switch the class type and expect to get a tensor grid with the exact same number of points unless you remember to change the input.

Of course, but there is no reason we need to do this.

> Even if you knew nothing about the deriving classes' inner workings, you would always know how many points you would get.

I can get the number of points for any point set by doing this...

```py
num_points = len(point_set.points)
```

Reply:
> I can get the number of points for any point set by doing this...
> num_points = len(point_set.points)

Yes, but you can't ask the `UnidimensionalPointSet` for a set with a desired number of points. You have to convert the number of points you want into a number corresponding to whatever parameter the class actually asks for. 

Having `UnidimensionalPointSet` ask for a number of points as a parameter is what I would prefer.

`ClenshawCurtisPointSet` does the same thing as `numpy.polynomial.chebyshev.chebpts2`, with the key difference being that `ClenshawCurtisPointSet` will output [0] while `chebpts2` will raise an error if you ask for less than 2 points. That function asks for number of points, and I can easily see someone new to this code expecting `ClenshawCurtisPointSet` to do the same thing.

What I don't understand is, why do you think the degree of the polynomial the points are the extrema of is so important that it should be the parameter the class asks for?

Reply:
Because doing so hides the math related to what level of approximation the points are derived from. It is straightforward to convert the degree of the Chebyshev polynomial or frequency of a trig function to a number of points, but going the other way around loses guarantees on having a complete set of points associated with a certain level of approximation. What I'm describing is also how these points would typically be constructed in 1d interpolation.

Reply:
> you can't ask the UnidimensionalPointSet for a set with a desired number of points

What is the use case you have in mind where this would be needed?

Reply:
> What I'm describing is also how these points would typically be constructed in 1d interpolation.

I looked up how scipy does it with the roots of the other quadratures, and I will concede that you are right. `ClenshawCurtisPointSet`will use degree.

(Interestingly, they all have the same number of roots as the degree of the polynomial. This explains why Clenshaw Curtis quadrate is the only one where the exponential growth is described by `2**x + 1` in the "1D Quadrature Rules for Sparse Grids" paper, while the rest use `2**(x+1) - 1`.)